gh-121710: Add PyBytesWriter API

vstinner opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-121710 -->
* Issue: gh-121710
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--121726.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

vstinner committed:
gh-121710: Add PyBytesWriter API

vstinner committed:
Reduce the small_buffer to 256 bytes

picnixz commented on Doc/c-api/bytes.rst:
```diff
@@ -201,3 +201,52 @@ called with a non-bytes parameter.
+   On success, update *\*str* and return ``0``.
+   On error, set an exception and return ``-1``.
+
+.. c:function:: int PyBytesWriter_WriteBytes(PyBytesWriter *writer, char **str, const void *bytes, Py_ssize_t size)
```
Do we need to always carry over \*\*str, or is it possible to have the *writer* instance holds a reference to it and be responsible for it (like for PyUnicode_Writer?) Or is the current interface not compatible with this kind of approach?

vstinner replied:
When I designed the private _PyBytesWriter API, I tried different approaches. This one ("carry str") is the most efficient: https://vstinner.github.io/pybyteswriter.html

picnixz replied:
Would it make sense to check if after (almost) 10 years this has been solved? I don't know when you checked the last time whether the machine code being produced is unoptimized, but are you interested in me trying to figure out something? Or do you want to keep the private API as it is?

picnixz replied:
So I checked for GCC 14 and Clang 18:

```C
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    char * restrict str;
    char  buffer[100];
} writer_t;

void init(writer_t *w) { w->str = w->buffer; }

static char* __attribute__((used)) 
f(int n)
{
    writer_t writer;
    init(&writer);
    while (1) {
        *(writer.str)++ = 'a';
    }
    *writer.str = '\0';
    return strdup(writer.buffer);
}

static char * __attribute__((used))
g(int n)
{
    writer_t writer;
    init(&writer);
    char *str = writer.str;
    while (1) {
        *str++ = 'a';
    }    
    *str = '\0';
    return strdup(writer.buffer);
}
```

I used `while(1)` just to force the compiler not to optimize using memset. The current version is `g` and the ideal one would be `f`. Here is the ASM output of gcc:

```asm
f:
        sub     rsp, 120
        mov     rdi, rsp
        call    init
.L4:    ; while (1) in f()
        mov     rax, QWORD PTR [rsp]  
        lea     rdx, [rax+1]		  
        mov     QWORD PTR [rsp], rdx  
        mov     BYTE PTR [rax], 97    
        jmp     .L4
g:
        sub     rsp, 120
        mov     rdi, rsp
        call    init
        mov     rax, QWORD PTR [rsp]
.L7:    ; while (1) in g()
        add     rax, 1
        mov     BYTE PTR [rax-1], 97
        jmp     .L7
```

So, yes, we still have the useless

```asm
mov     rax, QWORD PTR [rsp]  
lea     rdx, [rax+1]		  
mov     QWORD PTR [rsp], rdx  
mov     BYTE PTR [rax], 97
```

which would have been equivalent to

```asm
        mov     rax, QWORD PTR [rsp]
.L7:    ; while (1) in g()
        add     rax, 1
        mov     BYTE PTR [rax-1], 97
        jmp     .L7
```

And what's funny (it's not) is that using gcc 4.4.7 actually produces the following output:

```asm
init:
        lea     rax, [rdi+8]
        mov     QWORD PTR [rdi], rax
        ret
f:
        sub     rsp, 120
        mov     rdi, rsp
        call    init
        mov     rax, QWORD PTR [rsp]
.L4:
        mov     BYTE PTR [rax], 97
        add     rax, 1
        jmp     .L4
g:
        sub     rsp, 120
        mov     rdi, rsp
        call    init
        mov     rax, QWORD PTR [rsp]
.L8:
        mov     BYTE PTR [rax], 97
        add     rax, 1
        jmp     .L8
```

So I think we even regressed... Well, I agree that here it's an infinite loop so it's not really important but at least in v4, it was smart to put 

```asm
mov     rax, QWORD PTR [rsp]
``` 

outside of the loop... In the recent versions of GCC, it's not even the case =/

vstinner replied:
> So, yes, we still have the useless: (...)

I'm not surprised. You can compare by moving `char *str` outside the structure to check if the machine code looks more efficient.

picnixz commented on Modules/_testcapi/bytes.c:
```diff
@@ -37,8 +37,140 @@ bytes_resize(PyObject *Py_UNUSED(module), PyObject *args)
+static int
+bytes_equal(PyObject *obj, const char *str)
+{
+    return (PyBytes_Size(obj) == (Py_ssize_t)strlen(str)
+            && strcmp(PyBytes_AsString(obj), str) == 0);
```
```suggestion
    return (PyBytes_Size(obj) == (Py_ssize_t)strlen(str) &&
            strcmp(PyBytes_AsString(obj), str) == 0);
```

PEP 7

picnixz commented on Objects/bytesobject.c:
```diff
@@ -3640,6 +3730,25 @@ _PyBytesWriter_WriteBytes(_PyBytesWriter *writer, void *ptr,
+        return -1;
+    }
+
+    char *str2 = _PyBytesWriter_WriteBytes((_PyBytesWriter *)writer, *str,
```
In the other functions, you used `(_PyBytesWriter*)` (no space before `*`, so I think you should fix the other places (this place is PEP 7 compliant). I'll mark those that should be amended (maybe it's better that you do CTRL+R in your IDE since I could have missed some).

picnixz commented on Objects/bytesobject.c:
```diff
@@ -3379,12 +3379,45 @@ _PyBytesWriter_Init(_PyBytesWriter *writer)
+
+    char *str = _PyBytesWriter_Alloc(writer, size);
+    if (str == NULL) {
+        PyBytesWriter_Discard((PyBytesWriter*)writer);
```
```suggestion
        PyBytesWriter_Discard((PyBytesWriter *)writer);
```

picnixz commented on Objects/bytesobject.c:
```diff
@@ -3379,12 +3379,45 @@ _PyBytesWriter_Init(_PyBytesWriter *writer)
+    writer->overallocate = 1;
+
+    *pstr = str;
+    return (PyBytesWriter*)writer;
```
```suggestion
    return (PyBytesWriter *)writer;
```

picnixz commented on Objects/bytesobject.c:
```diff
@@ -3379,12 +3379,45 @@ _PyBytesWriter_Init(_PyBytesWriter *writer)
+void
+PyBytesWriter_Discard(PyBytesWriter *writer)
+{
+    _PyBytesWriter_Dealloc((_PyBytesWriter*)writer);
```
```suggestion
    _PyBytesWriter_Dealloc((_PyBytesWriter *)writer);
```

picnixz commented on Objects/bytesobject.c:
```diff
@@ -3449,6 +3482,27 @@ _PyBytesWriter_CheckConsistency(_PyBytesWriter *writer, char *str)
+        return -1;
+    }
+
+    _PyBytesWriter *writer = (_PyBytesWriter*)pub_writer;
```
```suggestion
    _PyBytesWriter *writer = (_PyBytesWriter *)pub_writer;
```

picnixz commented on Objects/bytesobject.c:
```diff
@@ -3546,6 +3600,27 @@ _PyBytesWriter_Prepare(_PyBytesWriter *writer, void *str, Py_ssize_t size)
+        return -1;
+    }
+
+    char *str2 = _PyBytesWriter_Prepare((_PyBytesWriter*)writer, *str, size);
```
```suggestion
    char *str2 = _PyBytesWriter_Prepare((_PyBytesWriter *)writer, *str, size);
```

picnixz commented on Objects/bytesobject.c:
```diff
@@ -3623,6 +3698,21 @@ _PyBytesWriter_Finish(_PyBytesWriter *writer, void *str)
+        return NULL;
+    }
+
+    PyObject *res = _PyBytesWriter_Finish((_PyBytesWriter*)writer, str);
```
```suggestion
    PyObject *res = _PyBytesWriter_Finish((_PyBytesWriter *)writer, str);
```

vstinner committed:
Fix _PyBytesWriter_Prepare()

Don't call _PyBytesWriter_Dealloc() on error.

vstinner committed:
Make _PyBytesWriter_Resize() static

vstinner committed:
Remove PyBytesWriter_CheckPtr()

The caller must check str. Any additional check has a negative impact
on performance.
